iothub/device/src/ModuleClient.cs

```diff
@@ -0,0 +1,511 @@
+
+
+        /// <summary>
+        /// Creates a ModuleClient instance in an IoT Edge deployment
+        /// based on environment variables.
+        /// </summary>
+        /// <param name="transportType">Specifies whether Amqp or Http transport is used</param>
+        /// <returns>ModuleClient instance</returns>
+        public static ModuleClient Create(TransportType transportType)
+        {
+            return Create(GetTransportSettings(transportType));
+        }
+
+        /// <summary>
+        /// Creates a ModuleClient instance in an IoT Edge deployment
+        /// based on environment variables.
+        /// </summary>
+        /// <param name="transportSettings">Prioritized list of transports and their settings</param>
+        /// <returns>ModuleClient instance</returns>
+        public static ModuleClient Create(ITransportSettings[] transportSettings)
+        {
+            return new EdgeModuleClientFactory(transportSettings).Create();
+        }
+
+        #region Module Specific API
+        /// <summary>
+        /// <param name="outputName">The output target for sending the given message</param>
+        /// <param name="message">The message to send</param>
+        /// <returns>The message containing the event</returns>
+        public Task SendEventAsync(string outputName, Message message)
+        {
+            // Codes_SRS_DEVICECLIENT_10_012: [If `outputName` is `null` or empty, an `ArgumentNullException` shall be thrown.]
+            if (string.IsNullOrWhiteSpace(outputName))
+            {
+                throw new ArgumentException(nameof(outputName));
+            }
+
+            // Codes_SRS_DEVICECLIENT_10_013: [If `message` is `null` or empty, an `ArgumentNullException` shall be thrown.]
+            if (message == null)
+            {
+                throw new ArgumentNullException(nameof(message));
+            }
+
+            // Codes_SRS_DEVICECLIENT_10_015: [The `output` property of a given `message` shall be assigned the value `outputName` before submitting each request to the transport layer.]
+            message.SystemProperties.Add(MessageSystemPropertyNames.OutputName, outputName);
+
+            // Codes_SRS_DEVICECLIENT_10_011: [The `SendEventAsync` operation shall retry sending `message` until the `BaseClient::RetryStrategy` tiemspan expires or unrecoverable error(authentication or quota exceed) occurs.]
+            return ApplyTimeout(operationTimeoutCancellationToken => this.InnerHandler.SendEventAsync(message, operationTimeoutCancellationToken));
+        }
+
+        /// <summary>
+        /// Sends a batch of events to device hub
+        /// <param name="outputName">The output target for sending the given message</param>
+        /// <param name="messages">A list of one or more messages to send</param>
+        /// </summary>
+        /// <returns>The task containing the event</returns>
+        public Task SendEventBatchAsync(string outputName, IEnumerable<Message> messages)
```

Comment:
>string outputName, IEnumerable<Message> messages [](start = 40, length = 48)

Consider swapping out the parameters to have "Message" first then Output. Here and elsewhere.

Reply:
Emm, this was how it was designed originally (by Olivier, etc.), and I am afraid this will have far reaching impact on current module developers. I am inclined to keep it as it is.